src/lib/markbind/src/preprocessors/componentPreprocessor.js

```diff
@@ -0,0 +1,394 @@
+const cheerio = require('cheerio');
+const nunjucks = require('nunjucks');
+const path = require('path');
+const url = require('url');
+
+const CyclicReferenceError = require('../handlers/cyclicReferenceError.js');
+
+const md = require('../lib/markdown-it');
+const utils = require('../utils');
+const urlUtils = require('../utils/urls');
+
+const _ = {};
+_.has = require('lodash/has');
+_.isEmpty = require('lodash/isEmpty');
+
+const {
+  ATTRIB_INCLUDE_PATH,
+  ATTRIB_CWF,
+} = require('../constants');
+
+function _preProcessAllComponents(node, context) {
+  const element = node;
+
+  // We do this since element.attribs is undefined if it does not exist
+  element.attribs = element.attribs || {};
+
+  element.attribs[ATTRIB_CWF] = path.resolve(context.cwf);
+}
+
+// TODO move this to componentParser
+function _preProcessThumbnail(node) {
```

Comment:
Should this be moved?

Reply:
Whether it should be moved or not will depend on how specialised componentProcessor will eventually get. I noted that there's some coupling between this and the parser that still needs to be resolved in future PRs, so I believe the answer will naturally come by as the refactoring process goes.

Also noted about the weird quirks that you have found. Refactoring by itself is usually quite a tough task by itself; verifying additional changes on top of it is going to make it even more tough, so always welcome to settle those problems in separate PRs (and good to open these in the issue tracker too).

Since my time is now quite restricted as compared to the past, I am no longer able to manually verify that everything works. So these green ticks approval that I give from now on would basically be just a glance through your code to see that there's no major source of issues in terms of design and implementation. You guys might want to test out each other's code from time to time to see if there's any problems.

With all that said, will be merging this. :+1: